From 12e34701f83624dd1c7962411901d41b3c0cca57 Mon Sep 17 00:00:00 2001
From: Johannes Berg <johannes@sipsolutions.net>
Date: Tue, 17 Nov 2020 19:55:50 +0100
Subject: [PATCH] _helpers: fix sign-compare warning

Fix two sign-compare warnings when this code is compiled on
(32-bit) platforms where ssize_t is 32 bits.

Signed-off-by: Johannes Berg <johannes@sipsolutions.net>
---
 lib/bup/_helpers.c | 8 +++++++-
 1 file changed, 7 insertions(+), 1 deletion(-)

diff --git a/lib/bup/_helpers.c b/lib/bup/_helpers.c
index e60dc5957c22..5a82e8cb6ad7 100644
--- a/lib/bup/_helpers.c
+++ b/lib/bup/_helpers.c
@@ -1068,7 +1068,13 @@ static PyObject *write_idx(PyObject *self, PyObject *args)
 	part = PyList_GET_ITEM(idx, i);
 	PyList_Sort(part);
 	plen = PyList_GET_SIZE(part);
-        if (plen > UINT32_MAX || UINT32_MAX - count < plen) {
+	/*
+	 * Limit to INT32_MAX so that on 32-bit platforms we don't get sign
+	 * comparison issues (and it's an unrealistic limit anyway); then
+	 * the cast to uint32_t is fine as it cannot lose/mess up anything
+	 * (considering also that plen as a list length cannot be negative.)
+	 */
+        if (plen > INT32_MAX || UINT32_MAX - count < (uint32_t)plen) {
             PyErr_Format(PyExc_OverflowError, "too many objects in index part");
             goto clean_and_return;
         }
-- 
2.26.2


